docs/source/howto/codes.rst

```diff
@@ -12,6 +12,34 @@ Interfacing external codes
 
 `#3986`_
 
+In order to work with an external simulation code in AiiDA, you need a calculation input plugin (subclassing the :py:class:`~aiida.engine.CalcJob` class) and an output parser plugin (subclassing the :py:class:`~aiida.parsers.Parser` class):
+
+Before starting to write a plugin, check on the `aiida plugin registry <https://aiidateam.github.io/aiida-registry/>`_ whether a plugin for your code is already available.
+
+Design guidelines
+------------------
+
+ * | **Start simple.**
+   | Make use of existing classes like :py:class:`~aiida.orm.nodes.data.dict.Dict`, :py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...
+   | Write only what is necessary to pass information from and to AiiDA.
+ * | **Don't break data provenance.**
+   | Store *at least* what is needed for full reproducibility.
```

Comment:
>  everything is needed for full reproducibility.

This is incorrect. For example, none of the outputs of a calculation need to be stored.

Also, this is not a guideline on how to run a study, it is a guideline on how to write a CalcJob plugin.

While "full reproducibility" is a (slightly too) strong word, it is really what AiiDA is about.
If your CalcJob plugin leaves out information that is needed for reproducing a calculation, then much of the benefit of AiiDA is lost.

I'm happy to reformulate it to make it more clear what I mean, e.g. something like
> Store *at least* what is needed for other to reproduce any calculation made with your plugin.

(open for suggestions!)

Reply:
Here is my reason for why the previous formulation is maybe not very helpful: Consider a plugin that interfaces with a code and passes only a single argument. As long as you keep track of the code version and the code runs deterministically, this calculation is fully reproducible. In fact you could even provide zero arguments if the code allows that and it would still be fully reproducible, i.e., it would produce the exact same result each time you run it.

So in that sense, the *only* thing that you need to provide for reproducible calculations are the arguments that you want to be able to vary as a plugin user. Am I missing something?

Reply:
Maybe the real point is to ensure that all input arguments are accurately tracked? So, for example, if I provide a filename as an argument, then of course it depends on the content of the file with that filename and just providing the filename would obviously not be sufficient?

Reply:
> Maybe the real point is to ensure that all input arguments are accurately tracked?

Yes, this is pretty much what I meant by storing the information needed to reproduce the calculations.
I'm very open to suggestions to clarify my original wording.